File: nodejs/patterns/src/creational/prototype.js
class Document {
  constructor(title, tags = []) {
    this.title = title;
    this.tags = tags;
  }

  clone() {
    return new Document(this.title, [...this.tags]);
  }
}

function demo() {
  const draft = new Document("Pattern Notes", ["draft"]);
  const copy = draft.clone();
  copy.tags.push("review");
  return draft.tags.length === 1 && copy.tags.length === 2;
}

module.exports = { Document, demo };

File: python/patterns/src/prototype_patterns/creational/prototype.py
"""Prototype pattern example."""

from __future__ import annotations

import copy
from dataclasses import dataclass, field


@dataclass
class Document:
    title: str
    tags: list[str] = field(default_factory=list)

    def clone(self) -> Document:
        return copy.deepcopy(self)


def demo() -> bool:
    draft = Document("Pattern Notes", ["draft"])
    clone = draft.clone()
    clone.tags.append("review")
    return draft.tags == ["draft"] and clone.tags == ["draft", "review"]
